python/ml4ir/base/model/cyclic_learning_rate.py

```diff
@@ -0,0 +1,281 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Cyclical Learning Rate Schedule policies for TensorFlow."""
+
+import tensorflow as tf
+#from tensorflow_addons.utils.types import FloatTensorLike
+
+#from typeguard import typechecked
+from typing import Union, Callable
+
+
+class CyclicalLearningRate(tf.keras.optimizers.schedules.LearningRateSchedule):
+    """A LearningRateSchedule that uses cyclical schedule."""
+
+    def __init__(
+        self,
+        initial_learning_rate,
+        maximal_learning_rate,
+        step_size,
+        scale_fn: Callable,
+        scale_mode: str = "cycle",
+        name: str = "CyclicalLearningRate",
+    ):
+        """Applies cyclical schedule to the learning rate.
+        See Cyclical Learning Rates for Training Neural Networks. https://arxiv.org/abs/1506.01186
+        ```python
+        lr_schedule = tf.keras.optimizers.schedules.CyclicalLearningRate(
+            initial_learning_rate=1e-4,
+            maximal_learning_rate=1e-2,
+            step_size=2000,
+            scale_fn=lambda x: 1.,
+            scale_mode="cycle",
+            name="MyCyclicScheduler")
+        model.compile(optimizer=tf.keras.optimizers.SGD(
+                                                    learning_rate=lr_schedule),
+                      loss='sparse_categorical_crossentropy',
+                      metrics=['accuracy'])
+        model.fit(data, labels, epochs=5)
+        ```
+        You can pass this schedule directly into a
+        `tf.keras.optimizers.Optimizer` as the learning rate.
+        Args:
```

Comment:
@mohazahran Can you follow the numpy docstring format? https://numpydoc.readthedocs.io/en/latest/format.html#sections